utils/include/detray/simulation/event_generator/random_track_generator.hpp

```diff
@@ -0,0 +1,306 @@
+
+    generator_t gen;
+    engine_t engine;
+
+    template <
+        typename T = generator_t,
+        std::enable_if_t<std::is_same_v<T, std::random_device>, bool> = true>
+    random_numbers() : gen{}, engine{gen()} {}
+
+    template <typename T = generator_t,
+              std::enable_if_t<std::is_same_v<T, std::seed_seq>, bool> = true>
+    random_numbers() : gen{42u}, engine{gen} {}
+
+    template <typename T = distribution_t,
+              std::enable_if_t<
+                  std::is_same_v<T, std::uniform_real_distribution<scalar_t>>,
+                  bool> = true>
+    DETRAY_HOST auto operator()(const scalar_t min, const scalar_t max) {
+        return distribution_t(min, max)(engine);
+    }
+
+    template <
+        typename T = distribution_t,
+        std::enable_if_t<std::is_same_v<T, std::normal_distribution<scalar_t>>,
+                         bool> = true>
+    DETRAY_HOST auto operator()(const scalar_t min, const scalar_t max) {
+        scalar_t mu{min + 0.5f * (max - min)};
+        return distribution_t(mu, 0.5f / 3.0f * (max - min))(engine);
+    }
+};
+
+/// @brief Generates track states with random momentum directions.
+///
+/// Generates the phi and theta angles of the track momentum according to a
+/// given random number distribution.
+///
+/// @tparam track_t the type of track parametrization that should be used.
+/// @tparam generator_t source of random numbers
+///
+/// @note Since the random number generator might not be copy constructible,
+/// neither is this generator. The iterators hold a reference to the rand
+/// generator, which must not be invalidated during the iteration.
+/// @note the random numbers are clamped to fit the phi/theta ranges. This can
+/// effect distribution mean etc.
+template <typename track_t, typename generator_t = random_numbers<>>
+class random_track_generator
+    : public detray::ranges::view_interface<random_track_generator<track_t>> {
+    public:
+    using point3 = __plugin::point3<detray::scalar>;
+    using vector3 = __plugin::vector3<detray::scalar>;
+
+    /// Configure how tracks are generated
+    struct configuration {
+
+        /// Gaussian vertex smearing
+        bool m_do_vtx_smearing = true;
+
+        /// How many tracks will be generated
+        std::size_t m_n_tracks{10u};
+
+        /// Range for theta and phi
+        std::array<scalar, 2> m_phi_range{-constant<scalar>::pi,
+                                          constant<scalar>::pi};
+        std::array<scalar, 2> m_theta_range{0.01f, constant<scalar>::pi};
+        /// Momentum range
+        std::array<scalar, 2> m_mom_range{1.f * unit<scalar>::GeV,
+                                          1.f * unit<scalar>::GeV};
+        /// Track origin
+        point3 m_origin{0.f, 0.f, 0.f}, m_origin_stddev{0.f, 0.f, 0.f};
+
+        /// Time parameter and charge of the track
+        scalar m_time{0.f * unit<scalar>::us}, m_charge{-1.f * unit<scalar>::e};
+
+        /// Setters
```

Comment:
Do you think we will need setter functions?

Reply:
Well, I think it would be nice to able to change the configuration in a clean way. I do that, for example, in the benchmark PR like this:
```
default_rkn_toygeo.config().n_barrel_layers(0u).n_endcap_layers(0u);
```